packages/jsii/lib/compiler.ts

```diff
@@ -72,28 +72,70 @@ export class Compiler implements Emitter {
   }
 
   /**
-     * Compiles the configured program.
-     *
-     * @param files can be specified to override the standard source code location logic. Useful for example when testing "negatives".
-     */
+   * Adds a listener for the 'result' event that is called with EmitResult.
+   *
+   * @param event    the 'result' event
+   * @param listener the listener to be called with the EmitResult.
+   */
+  public on(event: 'result', listener: (result: EmitResult) => any): this;
+  /**
+   * Adds a listener for the 'statusChanged' event that is called with a Diagnostic whenever the
+   * status of the current watch process changes.
+   *
+   * @param event    the 'statusChanged' event
+   * @param listener the listener to be called with the Diagnostic.
+   */
+  public on(event: 'statusChanged', listener: (message: Diagnostic) => any): this;
+  public on(event: string, listener: (...args: any[]) => any): this {
+    this.eventEmitter.on(event, listener);
+    return this;
+  }
+
+  /**
+   * Adds a once-only listener for the 'result' event that is called with EmitResult.
+   *
+   * @param event    the 'result' event
+   * @param listener the listener to be called with the EmitResult.
+   */
+  public once(event: 'result', listener: (result: EmitResult) => any): this;
+  /**
+   * Adds a once-only listener for the 'statusChanged' event that is called with a Diagnostic
+   * whenever the status of the current watch process changes.
+   *
+   * @param event    the 'statusChanged' event
+   * @param listener the listener to be called with the Diagnostic.
+   */
+  public once(event: 'statusChanged', listener: (message: Diagnostic) => any): this;
+  public once(event: string, listener: (...args: any[]) => any): this {
+    this.eventEmitter.once(event, listener);
+    return this;
+  }
+
+  /**
+   * Compiles the configured program.
+   *
+   * @param files can be specified to override the standard source code location logic. Useful for example when testing "negatives".
+   */
   public async emit(...files: string[]): Promise<EmitResult | never> {
     await this.buildTypeScriptConfig();
     await this.writeTypeScriptConfig();
     this.rootFiles = this.determineSources(files);
 
-    if (this.options.watch) {
-      if (files.length > 0) {
-        throw new Error('Files cannot be specified in watch mode!');
-      }
-      return this._startWatch();
-    }
     return this._buildOnce();
 
   }
 
+  public async watch(): Promise<Watch> {
+    await this.buildTypeScriptConfig();
+    await this.writeTypeScriptConfig();
+    this.rootFiles = this.determineSources([]);
```

Comment:
Can we reuse these three lines with `emit()`? `this.prepareForBuild()` or something